core/lib/zksync_core/src/commitment_generator/mod.rs

```diff
@@ -0,0 +1,230 @@
+use std::time::Duration;
+
+use anyhow::Context;
+use itertools::Itertools;
+use metrics::{CommitmentStage, METRICS};
+use multivm::zk_evm_latest::ethereum_types::U256;
+use tokio::sync::watch;
+use zksync_commitment_utils::{bootloader_initial_content_commitment, events_queue_commitment};
+use zksync_dal::{ConnectionPool, StorageProcessor};
+use zksync_types::{
+    commitment::{AuxCommitments, CommitmentCommonInput, CommitmentInput, L1BatchCommitment},
+    writes::{InitialStorageWrite, RepeatedStorageWrite, StateDiffRecord},
+    L1BatchNumber, ProtocolVersionId, StorageKey,
+};
+use zksync_utils::h256_to_u256;
+
+mod metrics;
+
+const SLEEP_INTERVAL: Duration = Duration::from_millis(100);
+
+#[derive(Debug)]
+pub struct CommitmentGenerator {
+    connection_pool: ConnectionPool,
+    stop_receiver: watch::Receiver<bool>,
+}
+
+impl CommitmentGenerator {
+    pub fn new(connection_pool: ConnectionPool, stop_receiver: watch::Receiver<bool>) -> Self {
+        Self {
+            connection_pool,
+            stop_receiver,
+        }
+    }
+
+    async fn calculate_aux_commitments(
+        conn: &mut StorageProcessor<'_>,
+        l1_batch_number: L1BatchNumber,
+        protocol_version: ProtocolVersionId,
+    ) -> anyhow::Result<AuxCommitments> {
+        let latency = METRICS.events_queue_commitment_latency.start();
+        let events_queue = conn
+            .blocks_dal()
+            .get_events_queue(l1_batch_number)
+            .await?
+            .context("Events queue is required for post-boojum batch")?;
+        let events_queue_commitment = events_queue_commitment(&events_queue, protocol_version)
+            .context("Events queue commitment is required for post-boojum batch")?;
+        latency.observe();
+
+        let latency = METRICS.bootloader_content_commitment_latency.start();
+        let initial_bootloader_contents = conn
+            .blocks_dal()
+            .get_initial_bootloader_heap(l1_batch_number)
+            .await?
+            .context("Bootloader initial heap is missing")?;
+        let bootloader_initial_content_commitment =
+            bootloader_initial_content_commitment(&initial_bootloader_contents, protocol_version)
+                .context("Bootloader content commitment is required for post-boojum batch")?;
```

Comment:
Given that calculating a commitment takes ~2s, it could make sense to refactor this method into 2 parts: loading data from Postgres (after which the DB connection is unused and may be dropped) and computations. Theoretically, computations could be put into `spawn_blocking()` task, but I don't see much benefit from it, TBH.

Reply:
I see value in it, I will also make these two (events and bootloader) commitments to be calculated concurrently